src/proper_fsm.erl

```diff
@@ -314,6 +367,108 @@ postcondition(#state{name = From, data = Data, mod = Mod}, Call, Res) ->
     Mod:postcondition(From, To, Data, Call, Res).
 
 
+%% -----------------------------------------------------------------------------
+%% Targeted command generation
+%% -----------------------------------------------------------------------------
+
+
+-spec targeted_commands_gen(mod_name()) -> proper_types:type().
+targeted_commands_gen(Mod) ->
+  State = initial_state(Mod),
+  ?LET([_ | Cmds], proper_statem:commands(?MODULE, State),
+       {finalize_weights(State, Cmds, maps:new()), Cmds}).
+
+-spec targeted_commands_gen(mod_name(), fsm_state()) -> proper_types:type().
+targeted_commands_gen(Mod, {Name, Data} = InitialState) ->
+  State = #state{name = Name, data = Data, mod = Mod},
+  ?LET([_ | Cmds], proper_statem:commands(?MODULE, State),
+       {finalize_weights(State, Cmds, maps:new()),
+        [{init, InitialState} | Cmds]}).
+
+-spec next_commands_gen(mod_name()) -> next_fun().
+next_commands_gen(Mod) ->
+  fun ({Weights, _Cmds}, {_D, T}) ->
+      State = initial_state(Mod),
+      NewWeights = proper_statem:next_weights(Weights, T),
+      CmdsGen = ?LET([_ | Cmds],
+                     proper_statem:next_gen(?MODULE, NewWeights, State),
+                     Cmds),
+      ?SHRINK(
+         ?LET(Cmds, ?LAZY(CmdsGen),
+              {finalize_weights(State, Cmds, NewWeights), Cmds}),
+         [CmdsGen])
+  end.
+
+-spec next_commands_gen(mod_name(), fsm_state()) -> next_fun().
+next_commands_gen(Mod, {Name, Data} = InitialState) ->
+  fun ({Weights, _Cmds}, {_D, T}) ->
+      State = #state{name = Name, data = Data, mod = Mod},
+      NewWeights = proper_statem:next_weights(Weights, T),
+      CmdsGen = ?LET([_ | Cmds],
+                     proper_statem:next_gen(?MODULE, NewWeights, State),
+                     [{init, InitialState} | Cmds]),
+      ?SHRINK(
+         ?LET(Cmds, ?LAZY(CmdsGen),
+              {finalize_weights(State, Cmds, NewWeights), Cmds}),
+         [CmdsGen])
+  end.
+
+%% Commands are selected according to the weights found in the
+%% map provided, or from a `weight/3' callback.
+
+%% @private
+-spec select_command(state(), weights()) -> proper_types:type().
+select_command(State, Weights) ->
+  #state{name = From, data = Data, mod = Mod} = State,
+  SelectAux =
+    fun ({To, {call, _Mod, Call, _Args} = SymbCall}) ->
+        RealTo = cook_history(From, To),
+        Key = {From, RealTo, Call},
+        case maps:is_key(Key, Weights) of
```

Comment:
I think this `case` is better if it computes/returns the `Weight` element, with the `{Weight, SymbCall}` 2-tuple construction being in its own line below the `case`.